docs: reorganize documentation, add new guides (#17)
threeport/user-docs: docs/architecture/control-plane.md

These are capable tools for certain
use cases.  However, using the AWS SDK directly gives us the most flexibility
and ensures we don't encounter any unsupported operations we might need to
perform in managing cloud resources for Threeport users.  It also serves as a
reference implementation for platform engineers that extend Threeport and wish
to use a similar approach.

### Control Plane Controller

The control plane controller allows users of Threeport to deploy new Threeport
control planes using Threeport itself.  This is available so that large organization
that wish to clearly separate concerns between lines of business can do so in a
Threeport-native way without the need to use `tptctl` to bootstrap new control
planes when needed.  It also allows separation of tiers, e.g. development and
production if desired.  However, we don't recommend this approach for most users
unless they have a compelling need for it.

### Gateway Controller

The gateway controller manages network ingress support services when a workload
has such a dependency.

The following support services are installed on Kubernetes as needed by the
Gateway controller:

> * [Gloo Gateway](https://github.com/solo-io/gloo): the network ingress

Reviewer: ```suggestion
* [Gloo Edge](https://github.com/solo-io/gloo): the network ingress
```

See banner in Gloo Edge docs... https://docs.solo.io/gloo-edge/latest/

controller used by Threeport.
* [cert-manager](https://github.com/cert-manager/cert-manager): used to
provision and rotate TLS certificates.
* [external-dns](https://github.com/kubernetes-sigs/external-dns): manages DNS
records created for workloads.

When a support service controller needs to be installed in Kubernetes, we use
the
[support-services-operator](https://github.com/nukleros/support-services-operator)
to perform the install.  The Kubernetes manifest provided to the Workload is
actually a custom resource that is managed by the support-services-operator.  It
installs the support services listed above in this manner.

In addition to the support services installations on Kubernetes, the gateway
controller appends Kubernetes resources to those defined by the user with the
Workload resource to configure the support service for that workload.

### Helm Workload Controller

> The helm workload controller uses the popular Kubernetes package manager,

Reviewer: I think we should capitalize `Helm` and `Terraform` everywhere, unless it's used in something like a variable name and wouldn't make sense.

Author: Done

[Helm](https://helm.sh/) to deploy workloads in Kubernetes.  Helm templates have
drawbacks in complex environments since templating is inherently inferior to
general purpose programming languages.  More on this topic is discussed in the
[Continuous Delivery & GitOps
section](../../concepts/application-orchestration#continuous-delivery-gitops)
of the Application Orchestration document in another section of these Threeport
docs.  We much prefer Go programs to construct Kubernetes resources, however,
when Helm charts are available for open source projects and when they are in
extensive use by a team, it is valuable to have the option available.  We also

> currently use it to implement observability systems.

Reviewer: ```suggestion
of the Application Orchestration document.  We prefer Go programs to construct
Kubernetes resources.  However, Helm support is still valuable because many open-source
charts are available and it is already in extensive use by many teams.

Although we prefer Go to manage Kubernetes resources, we recognize there are
use-cases where it is more appropriate to use Helm. Threeport's implementation
of the [Observability Stack](../../observability/observability-intro/) is one
example.  The requirements for observability line up well with what is already
provided by open-source Helm charts, so it made more sense to implement this
controller with Threeport's Helm integration.
```

Slight rephrase and example of Helm implementation in Threeport.

### Kubernetes Runtime Controller

The Kubernetes runtime controller is used to provision new Kubernetes
environments for workloads.  It serves as an infra provider agnostic abstraction
that allows a user to provision environments with the infra provider as a simple
attribute of the `KubernetesRuntimeDefinition` object.  The
`KubernetesRuntimeInstance` object is where the connection information for each

> cluster API is stored and utilized when workloads are deployed to that cluster.

Reviewer: ```suggestion
`KubernetesRuntimeInstance` object is where connection information for each
cluster is stored and utilized when workloads are deployed to that cluster.
```

Wanted to avoid the term "cluster API" since that's the name of a project that we've chosen not to use.

### Observability Controller

The observability controller is responsible for deploying observability stacks
when a Threeport user wishes to have access to metrics and logs from their
workloads.

The following components are available to deploy as a part of the stack:

* [Prometheus](https://prometheus.io/docs/introduction/overview/): the metrics
  collection system.
* [Promtail](https://grafana.com/docs/loki/latest/send-data/promtail/): log
  forwarding from individual machines to the log storage back end.
* [Loki](https://github.com/grafana/loki): log storage.
* [Grafana](https://github.com/grafana/grafana): observability dashboard.

The observability controller leverages the helm workload controller to install

> helm charts to deploy each of these components.

Reviewer: ```suggestion
The observability controller leverages the Helm workload controller to install
Helm charts to deploy each of these components.
```

Author: Updated separately

provision custom infrastructure needed by workloads.  Terraform is less than
ideal for provisioning infrastructure in a control plane like Threeport for
reasons discussed [elsewhere](../../concepts/application-orchestration#continuous-delivery-gitops)
but it is offered in Threeport for two reasons.  Many teams have made extensive
use of Terraform and this allows them to use those configs in Threeport.  Also,
Threeport offers native support for only a small number of AWS managed services
and Terraform offers support for a much larger number of those
resources.  Using Terraform for simpler use cases can be useful compared to the
alternative of developing a custom Threeport extension to manage those
same AWS resources.

> Note: Terraform is only supported for managing AWS resource at this time.

### Workload Controller

The workload controller deploys a defined set of Kubernetes resources to a
nominated (or default) Kubernetes runtime instance.  This controller is quite
rudimentary in that the user is required to define the granular detail of all
Kubernetes resources that constitute their workload.  However, it is useful in

> simple implementations and, when paired with a Kubernetes Operator or a custom

Reviewer: ```suggestion
simple implementations.  When paired with a Kubernetes Operator or a custom
```
